internal/files/writer.go

```diff
@@ -201,15 +202,34 @@ func HardlinkDir(olddir, newdir string) error {
 		}
 	}
 
-	return filepath.Walk(olddir, func(oldpath string, info fs.FileInfo, err error) error {
+	rootInfo, err := os.Lstat(olddir)
+	if err != nil {
+		return fmt.Errorf("cannot stat olddir %v: %w", olddir, err)
+	}
+
+	err = os.MkdirAll(newdir, rootInfo.Mode())
+	if err != nil {
+		return fmt.Errorf("cannot create new root dir %v: %w", olddir, err)
+	}
+
+	fastwalkConf := fastwalk.DefaultConfig.Copy()
+	fastwalkConf.Sort = fastwalk.SortDirsFirst
+
+	return fastwalk.Walk(fastwalkConf, olddir, func(oldpath string, d os.DirEntry, err error) error {
 		if err != nil {
-			return fmt.Errorf("failed to walk dir: %v, %w", info, err)
+			return fmt.Errorf("failed to walk dir: %v, %w", oldpath, err)
 		}
 
 		newpath := filepath.Join(newdir, strings.TrimPrefix(oldpath, olddir))
 
-		if info.IsDir() {
-			err := os.MkdirAll(newpath, info.Mode())
+		if d.IsDir() {
+			info, err := d.Info()
+
+			if err != nil {
+				return fmt.Errorf("unable to get directory info %v: %w", oldpath, err)
+			}
+
+			err = os.MkdirAll(newpath, info.Mode())
```

Comment:
This is like `mkdir -p` right, which is probs slower than just a straight mkdir cause it has to stat all the parents every time? I think it was likely being defensive against the parent dirs not existing yet, but knowing what we know now, I think it would be safe to create the root newdir once before starting the walk, and then making this just do a straight mkdir?

Also, does `info.Mode()` do a separate stat syscall to get its info? Any way to avoid that if so?

Reply:
For `info.Mode()` it looks like it doesn't do extra syscalls looking at the implementations. It's a 64bit int that is masked to indicate the permissions. We could just copy and use the the `rootDir`'s info and pass that into the walkDir. I assume that the child directories can have the same permissions as the parent.

I don't know enough about the file structure, would there be a case where a child directory would be less permissive than the parent or more permissive? Seems like it would be pretty safe to just use the root's permissions for the directory and save on the costs there. Using the root would save us on the `lstat` call in `dirEntry.info()`

Reply:
I think that is maybe fine, but the risk of it not being fine is not worth it IMO. I am not super familiar with how picky `npm` and/or `yarn` are about the permissions on the various files and folders they manage, but I do know at least some of the bits need their executable bit set, and so I'd vote for maintaining the interface we had before. if we still see this as a big bottleneck later i think we could do some proving that the perms are fine to just set to consistent values, or we could have the DL server ship us the exact perms of all the stuff alongside the tarball so we can just have it in memory.

Reply:
The file permissions should be the same since we're hard linking them, this is only the dir creation. Likely dir creation is much less frequent than files. But better safe than sorry.